doc/design/prune.md

```diff
@@ -0,0 +1,107 @@
+# Prune
+
+## Motivation
+
+We want to support running inference, training and checkpointing in one `ProgramDesc`. We implement 
+`void Prune(const ProgramDesc* input, ProgramDesc* output, int id)` function, which takes a `ProgramDesc`
+and generate a pruned `ProgramDesc`.
+
+## Challenge
+
+Pruning need to support both variables and operators being evaluation targets. Consider the following
+different situations.
+
+```python
+# Case 1: run foward pass.
+cost_np = session.run(target=cost)
+# Case 2: run backward passing.
+opts_np, _ = session.run(target=[cost, opt])
+# Case 3: run checkpointing
+_ = session.run(target=checkpoint)
+```
+
+## Solution
+
+To support evaluation of operators, we add `is_target` field in the `OpDesc`.
+
+```c++
+message OpDesc {
+  required string type = 3;
+  repeated Var inputs = 1;
+  repeated Var outputs = 2;
+  repeated Attr attrs = 4;
+  required bool is_target = 5 [ default = false ];
+};
+```
+
+To support evaluation of variables, we add [fetch_op](https://github.com/PaddlePaddle/Paddle/pull/4599).
+For each variable in the `target`, we insert a `fetch_op` into the `ProgramDesc` with `variable` being
+`fetch_op`'s input. Then we also set `fetch_op` is a target.
+
+### Algorithm
+
+If an operator needs to be run, it must fall into one of the following cases:
+
+1. It is the target.
+2. It is depended by some other ops, meaning its output is some other op's input.
+
+The first case can be checked by `op_desc.is_traget()` . The second case can be implement as
+
+```c++
+bool HasDependentVar(const OpDesc& op_desc, const std::set<string>& dependent_vars) {
+  for (auto& var : op_desc.outputs()) {
+    for (auto& argu : var.arguments()) {
+      if (dependent_vars.count(argu) != 0) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+```
+
+
+
+Then the whole algorithm can be implemented as the following
+
+```c++
+void Prune(const ProgramDesc& input, ProgramDesc* output, int id) {
```

Comment:
What is the purpose of having `id` as an argument for `Prune`? How should we use the `Prune` function with the `id` argument?

Reply:
`id` is the block_id. Just like `executor::run` takes a block_id, it denotes which block, usually the root block, to be pruned.
I will change the name to `block_id` to make it less confusing.

Reply:
I thought `Prune` will always take the *entire* given `ProgramDesc`, and outputs a minimal `ProgramDesc`. If we add `block_id = 2`, what should `Prune` do to other blocks (e.g., block 0, block 1): keep it or discard it / prune it or not?

Reply:
I want `Prune` to be called recursively. Say it encounters a `RNNOp` in block 0, then we can simply call `prune(rnn_sub_block_id)` to do the pruning there. Although the support for RNN and IfElse block is still under development, I believe we should leave the `block_id` as an argument here.

Reply:
One alternative to consider is we can have something like this:

// private
void pruneImpl(const ProgramDesc& input, ProgramDesc* output, int id) {
    if (rnn) {
        pruneImpl(input, output, id+1);
        return;
    }
}

// public
void Prune(const ProgramDesc& input, ProgramDesc* output) {
    pruneImpl(input, output, 0)
}

Reply:
I can't agree more. It is awesome.

Reply:
Not relevant to this PR, I think same design concept could be applied to `executor` as well.

Reply:
Good idea! If you want to send a PR changing the executor design doc, I will instantly approve :p